rfcs/999-kustomize.adoc

```diff
@@ -0,0 +1,197 @@
+centrally supported sidecars, or completely bespoke ones) should be present.
+Today, that's supported by a Helm chart specified by `.spec.template.chart`,
+and tweaks to the configuration are injected as values through
+`.spec.template.values`, both in the Application object.
+
+Using Kustomize, the Application object would lose both `.spec.template.chart`
+and `.spec.template.values`, having those replaced by a single
+`.spec.template.kustomization`, as exemplified below, containing an inline
+Kustomization object.
+
+.application.yaml
+[source,yaml]
+----
+apiVersion: shipper.booking.com/v1alpha1
+kind: Application
+metadata:
+  name: bikerental
+spec:
+  template:
+    clusterRequirements:
+      regions:
+      - name: eu-nl
+      capabilities: []
+    kustomization:
+      image:
+        name: juliogreff/bikerental
+        tag: deadcafe
+      bases:
+        - https://example.com/core
+      namePrefix: bikerental-
+      commonLabels:
+        app: bikerental
+      patchesStrategicMerge:
+        - https://example.com/sidecars/sidecar.yaml
+    strategy:
+      steps:
+      - name: full on
+        capacity:
+          incumbent: 0
+          contender: 100
+        traffic:
+          incumbent: 0
+          contender: 100
+----
+
+Note the URL `.spec.template.kustomization` points to. That's the application's
+configuration bundle. As the `kustomize` CLI tool, Shipper will also look for a
+`kustomization.yaml` inside that path.
+
+.https://example.com/sidecars/sidecar.yaml
+[source,yaml]
+----
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: app # <1>
+spec:
+  template:
+    spec:
+      containers:
+        - name: sidecar
+          image: sidecar
+----
+<1> There's a very important downside to be noted, in which sidecar patches
+must know the name of the deployment they're patching (minus any prefixes and
+suffixes), so proper conventions will be essential.
+
+.https://example.com/core/kustomization.yaml
+[source,yaml]
+----
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+  - deployment.yaml
+  - service.yaml   # <1>
+  - configmap.yaml # <1>
+----
+
+.https://example.com/core/deployment.yaml
+[source,yaml]
+----
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: app
+spec:
+  replicas: 1
+  template:
+    spec:
+      containers:
+      - name: app
+        image: app
+        ports:
+          - containerPort: 80
+----
+<1> Kustomizations can include as many of any Kubernetes resources as they
+need. As those are not of consequence to this RFC, they are omitted for
+brevity.
+
+The kustomizations above cover the application level configuration. At release
+time, Shipper still needs to apply its own little bit of configuration with
+data from the Release object it just created. That kustomization would be the
+equivalent of the following YAML snippet. It injects references to the release
+hash through `nameSuffix` and `commonLabels`, adjusts the amount of replicas
```

Comment:
The `commonLabels` functionality is nice. We might be able to lean on this to clean up some of the gnarly object mutation bits in the Installation controller.